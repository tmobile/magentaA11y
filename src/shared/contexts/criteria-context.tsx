import {
  ContentTab,
  SavedCriteria,
} from 'components/content-display/markdown-content/markdown-content.types';
import React, { createContext, ReactNode, useContext } from 'react';

const LOCAL_STORAGE_KEY: string = 'savedCriteria';

interface CriteriaContextType {
  savedCriteria: SavedCriteria[];
  saveCriteria: (criteria: SavedCriteria) => void;
  removeCriteria: (id: string) => void;
  clearCriteria: (label: string) => void;
}

const CriteriaContext = createContext<CriteriaContextType | undefined>(
  undefined
);

const saveToLocalStorage = (criteria: SavedCriteria[]) => {
  try {
    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(criteria));
  } catch (error) {
    console.error('Failed to save criteria to localStorage:', error);
  }
};

const getInitialSavedCriteria = (): SavedCriteria[] => {
  try {
    const storedCriteria = localStorage.getItem(LOCAL_STORAGE_KEY);
    return storedCriteria ? JSON.parse(storedCriteria) : [];
  } catch (error) {
    console.error('Failed to load criteria from localStorage:', error);
    return [];
  }
};

const useSavedCriteria = () => {
  const [savedCriteria, setSavedCriteria] = React.useState<SavedCriteria[]>(
    getInitialSavedCriteria
  );

  const saveCriteria = React.useCallback((criteria: SavedCriteria) => {
    setSavedCriteria((prev) => {
      if (
        !prev.some(
          (item) =>
            item.label === criteria.label && item.content === criteria.content
        )
      ) {
        const updatedCriteria = [...prev, criteria];
        saveToLocalStorage(updatedCriteria);
        return updatedCriteria;
      }
      return prev;
    });
  }, []);

  const removeCriteria = React.useCallback((id: string) => {
    setSavedCriteria((prev) => {
      const updatedCriteria = prev.filter((item) => item.id !== id);
      saveToLocalStorage(updatedCriteria);
      return updatedCriteria;
    });
  }, []);

  const clearCriteria = React.useCallback((criteria: string) => {
    setSavedCriteria((prev) => {
      const updatedCriteria = prev.filter(
        (item: ContentTab) => item.tab !== criteria
      );
      saveToLocalStorage(updatedCriteria);
      return updatedCriteria;
    });
  }, []);

  return {
    savedCriteria,
    saveCriteria,
    removeCriteria,
    clearCriteria,
  };
};

export const CriteriaProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const criteria = useSavedCriteria();

  return (
    <CriteriaContext.Provider value={criteria}>
      {children}
    </CriteriaContext.Provider>
  );
};

export const useCriteria = () => {
  const context = useContext(CriteriaContext);
  if (!context) {
    throw new Error('useCriteria must be used within a CriteriaProvider');
  }
  return context;
};
