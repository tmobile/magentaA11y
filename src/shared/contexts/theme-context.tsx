import React, { createContext, useContext, useEffect, useState } from "react";

export enum Theme {
  LIGHT = "light",
  DARK = "dark",
}

// Types for the context
interface ThemeContextType {
  theme: Theme;
  setTheme: (theme: Theme) => void;
}

// Default context value
const ThemeContext = createContext<ThemeContextType>({
  theme: Theme.LIGHT,
  setTheme: () => {},
});

// Provider component
export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [theme, setTheme] = useState<Theme>(() => {
    const savedTheme = localStorage.getItem("theme");
    return savedTheme === Theme.DARK || savedTheme === Theme.LIGHT
      ? (savedTheme as Theme)
      : window.matchMedia("(prefers-color-scheme: dark)").matches
      ? Theme.DARK
      : Theme.LIGHT;
  });

  // Apply the theme to the document
  useEffect(() => {
    document.documentElement.setAttribute("data-theme", theme);
  }, [theme]);

  // Listen for system theme changes
  useEffect(() => {
    const systemThemeChangeListener = (e: MediaQueryListEvent) => {
      const systemTheme = e.matches ? Theme.DARK : Theme.LIGHT;
      const savedTheme = localStorage.getItem("theme");

      // Only update theme if user hasn't manually overridden
      if (!savedTheme) {
        setTheme(systemTheme);
      }
    };

    const darkModeMediaQuery = window.matchMedia(
      "(prefers-color-scheme: dark)"
    );
    darkModeMediaQuery.addEventListener("change", systemThemeChangeListener);

    // Cleanup listener on unmount
    return () => {
      darkModeMediaQuery.removeEventListener(
        "change",
        systemThemeChangeListener
      );
    };
  }, []);

  // Update localStorage only when the user explicitly changes the theme
  const handleSetTheme = (newTheme: Theme) => {
    setTheme(newTheme);
    localStorage.setItem("theme", newTheme);
  };

  return (
    <ThemeContext.Provider value={{ theme, setTheme: handleSetTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

// Custom hook for accessing the theme context
export const useTheme = () => useContext(ThemeContext);
