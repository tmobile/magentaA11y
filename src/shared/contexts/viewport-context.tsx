import { createContext, useContext, useEffect, useState } from "react";

interface ViewportContextProps {
  isMobile: boolean;
  isLargeTablet: boolean;
}

const ViewportContext = createContext<ViewportContextProps | undefined>(
  undefined
);

export const ViewportProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [isMobile, setIsMobile] = useState(false);
  const [isLargeTablet, setIsLargeTablet] = useState(false);

  useEffect(() => {
    const mobileMediaQuery = window.matchMedia("(max-width: 768px)");
    const tabletMediaQuery = window.matchMedia("(max-width: 936px)");

    setIsMobile(mobileMediaQuery.matches);
    setIsLargeTablet(tabletMediaQuery.matches);

    const handleMobileResize = () => setIsMobile(mobileMediaQuery.matches);
    const handleTableteResize = () => setIsLargeTablet(tabletMediaQuery.matches);
    mobileMediaQuery.addEventListener("change", handleMobileResize);
    tabletMediaQuery.addEventListener("change", handleTableteResize);

    return () => {
      mobileMediaQuery.removeEventListener("change", handleMobileResize);
      tabletMediaQuery.removeEventListener("change", handleTableteResize);
    };
  }, []);

  return (
    <ViewportContext.Provider value={{ isMobile, isLargeTablet }}>
      {children}
    </ViewportContext.Provider>
  );
};

export const useViewport = (): ViewportContextProps => {
  const context = useContext(ViewportContext);
  if (!context) {
    throw new Error("useViewport must be used within a ViewportProvider");
  }
  return context;
};
