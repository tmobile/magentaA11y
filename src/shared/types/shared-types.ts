export enum DocumentationCategory {
  WEB = 'web',
  NATIVE = 'native',
  HOW_TO_TEST = 'how-to-test',
}
