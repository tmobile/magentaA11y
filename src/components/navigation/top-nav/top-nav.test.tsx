import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { ViewportProvider } from 'shared/contexts/viewport-context';
import { DocumentationCategory } from 'shared/types/shared-types';
import { TopNavProps } from '../nav.types';
import TopNav, { getFirstOverviewLink } from './top-nav';

const renderWithProviders = (
  ui: React.ReactElement,
  { initialEntries = ['/home'] } = {}
) => {
  return render(
    <ViewportProvider>
      <MemoryRouter
        initialEntries={initialEntries}
        future={{ v7_relativeSplatPath: true, v7_startTransition: true }}>
        {ui}
      </MemoryRouter>
    </ViewportProvider>
  );
};

const mockNavItems: TopNavProps['navItems'] = [
  { label: 'Web Criteria', href: '/web-criteria' },
  { label: 'Native App Criteria', href: '/native-criteria' },
  { label: 'About Us', href: '/about' },
];

describe('TopNav Component - Snapshot Test', () => {
  test('matches the snapshot', () => {
    const { asFragment } = renderWithProviders(
      <TopNav navItems={mockNavItems} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('TopNav Component - Basic Rendering', () => {
  test('renders the TopNav component', () => {
    renderWithProviders(<TopNav navItems={mockNavItems} />);
    expect(screen.getByRole('navigation')).toBeInTheDocument();
  });

  test('contains the correct class name', () => {
    renderWithProviders(<TopNav navItems={mockNavItems} />);
    const nav = screen.getByRole('navigation');
    expect(nav).toHaveClass('MagentaA11y__navbar__nav');
  });

  test('navigation list exists', () => {
    renderWithProviders(<TopNav navItems={mockNavItems} />);
    const navList = screen.getByRole('list');
    expect(navList).toBeInTheDocument();
  });
});

describe('TopNav Component - Content & Accessibility Tests', () => {
  test('each NavLink has the correct label', () => {
    renderWithProviders(<TopNav navItems={mockNavItems} />);

    mockNavItems.forEach((item) => {
      const matchingLinks = screen.getAllByRole('link', {
        name: item.ariaLabel,
      });
      expect(matchingLinks.length).toBeGreaterThan(0);
    });
  });

  test('nav items render with correct labels', () => {
    renderWithProviders(<TopNav navItems={mockNavItems} />);

    mockNavItems.forEach((item) => {
      expect(screen.getByText(item.label)).toBeInTheDocument();
    });
  });

  test('menu button has correct accessible labels', () => {
    window.matchMedia = jest.fn().mockImplementation((query) => ({
      matches: query === '(max-width: 936px)',
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    }));

    renderWithProviders(<TopNav navItems={mockNavItems} />);

    const menuButton = screen.getByRole('button');
    expect(menuButton).toHaveAttribute(
      'aria-label',
      expect.stringMatching(/menu|close/i)
    );
  });

  test('navigation section has the correct ARIA label', () => {
    renderWithProviders(<TopNav navItems={mockNavItems} />);

    const nav = screen.getByRole('navigation');
    expect(nav).toHaveAttribute('aria-label', 'main');
  });
});

describe('TopNav Component - Interaction Tests', () => {
  beforeEach(() => {
    // Mock matchMedia for mobile behavior
    window.matchMedia = jest.fn().mockImplementation((query) => ({
      matches: query === '(max-width: 936px)',
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    }));
  });

  test('clicking the menu button toggles expanded state (Mobile Only)', async () => {
    renderWithProviders(<TopNav navItems={mockNavItems} />);
    const menuButton = screen.getByRole('button');

    // Ensure menu starts closed
    expect(menuButton).toHaveAttribute('aria-expanded', 'false');

    // Click to open menu
    await userEvent.click(menuButton);
    expect(menuButton).toHaveAttribute('aria-expanded', 'true');

    // Click to close menu
    await userEvent.click(menuButton);
    expect(menuButton).toHaveAttribute('aria-expanded', 'false');
  });

  test('clicking a navigation item closes the menu (Mobile Only)', async () => {
    renderWithProviders(<TopNav navItems={mockNavItems} />);
    const menuButton = screen.getByRole('button');

    // Open the menu
    await userEvent.click(menuButton);
    expect(menuButton).toHaveAttribute('aria-expanded', 'true');

    // Click on a nav link
    const navLink = screen.getByRole('link', { name: /web criteria/i });
    await userEvent.click(navLink);

    // Menu should be closed
    expect(menuButton).toHaveAttribute('aria-expanded', 'false');
  });

  test('clicking outside the nav closes the menu (Mobile Only)', async () => {
    renderWithProviders(<TopNav navItems={mockNavItems} />);
    const menuButton = screen.getByRole('button');

    // Open the menu
    await userEvent.click(menuButton);
    expect(menuButton).toHaveAttribute('aria-expanded', 'true');

    // Simulate clicking outside the menu
    fireEvent.mouseDown(document.body);

    // Menu should be closed
    expect(menuButton).toHaveAttribute('aria-expanded', 'false');
  });

  test('handles keyboard navigation correctly (Mobile Only)', async () => {
    renderWithProviders(<TopNav navItems={mockNavItems} />);
    const menuButton = screen.getByRole('button');

    // Open the menu
    await userEvent.click(menuButton);
    expect(menuButton).toHaveAttribute('aria-expanded', 'true');

    // Get focusable elements inside the menu
    const focusableElements = screen.getAllByRole('link');
    expect(focusableElements.length).toBeGreaterThan(0);

    let currentIndex = 1; // setting to 1 to skip the home link

    // Focus the first element and simulate Tab key press
    await userEvent.tab();
    expect(focusableElements[currentIndex]).toHaveFocus();

    // Press Tab to move forward
    currentIndex++;
    await userEvent.tab();
    expect(focusableElements[currentIndex]).toHaveFocus();

    // Shift + Tab should move backward
    currentIndex--;
    await userEvent.keyboard('{Shift>}{Tab}{/Shift}');
    expect(focusableElements[currentIndex]).toHaveFocus();

    // Simulate pressing Tab on last element
    for (let i = currentIndex; i < focusableElements.length; i++) {
      await userEvent.tab();
    }

    // Pressing Tab again should close the menu
    expect(menuButton).toHaveAttribute('aria-expanded', 'false');
  });
});

describe('TopNav Component - Navigation Behavior Tests', () => {
  test('each NavLink has the correct href attribute', () => {
    renderWithProviders(<TopNav navItems={mockNavItems} />);

    expect(screen.getByRole('link', { name: 'Web Criteria' })).toHaveAttribute(
      'href',
      getFirstOverviewLink(DocumentationCategory.WEB)
    );

    expect(
      screen.getByRole('link', { name: 'Native App Criteria' })
    ).toHaveAttribute(
      'href',
      getFirstOverviewLink(DocumentationCategory.NATIVE)
    );

    expect(screen.getByRole('link', { name: 'About Us' })).toHaveAttribute(
      'href',
      '/about'
    );
  });

  test('clicking a NavLink updates the URL', async () => {
    const user = userEvent.setup();

    renderWithProviders(
      <>
        <TopNav navItems={mockNavItems} />
        <Routes>
          <Route
            path="/web-criteria/*"
            element={<div data-testid="web-page">Web Criteria Page</div>}
          />
          <Route
            path="/native-criteria/*"
            element={<div data-testid="native-page">Native Criteria Page</div>}
          />
          <Route
            path="/about"
            element={<div data-testid="about-page">About Page</div>}
          />
        </Routes>
      </>,
      { initialEntries: ['/'] }
    );

    await user.click(screen.getByRole('link', { name: 'Web Criteria' }));
    expect(await screen.findByTestId('web-page')).toBeInTheDocument();

    await screen.findByRole('link', { name: 'Native App Criteria' });

    await user.click(screen.getByRole('link', { name: 'Native App Criteria' }));
    expect(await screen.findByTestId('native-page')).toBeInTheDocument();

    await screen.findByRole('link', { name: 'About Us' });
    await user.click(screen.getByRole('link', { name: 'About Us' }));
    expect(await screen.findByTestId('about-page')).toBeInTheDocument();
  });

  test('active class is applied to the correct link', async () => {
    const user = userEvent.setup();

    renderWithProviders(
      <Routes>
        <Route path="*" element={<TopNav navItems={mockNavItems} />} />
      </Routes>,
      { initialEntries: ['/web-criteria'] }
    );

    expect(screen.getByRole('link', { name: 'Web Criteria' })).toHaveClass(
      'active'
    );

    await user.click(screen.getByRole('link', { name: 'Native App Criteria' }));
    expect(screen.getByRole('link', { name: 'Native App Criteria' })).toHaveClass(
      'active'
    );

    await user.click(screen.getByRole('link', { name: 'About Us' }));
    expect(screen.getByRole('link', { name: 'About Us' })).toHaveClass(
      'active'
    );
  });
});

test('menu button does not render on desktop view', () => {
  window.matchMedia = jest.fn().mockImplementation((query) => ({
    matches: query !== '(max-width: 936px)',
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  }));

  renderWithProviders(<TopNav navItems={mockNavItems} />);

  expect(screen.queryByRole('button')).not.toBeInTheDocument();
});
