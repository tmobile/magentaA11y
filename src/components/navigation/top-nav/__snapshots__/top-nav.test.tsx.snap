// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopNav Component - Snapshot Test matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MagentaA11y__navbar"
    data-theme="dark"
  >
    <div
      class="MagentaA11y__brand"
    >
      <a
        aria-current="page"
        aria-label="Magenta A11y - Home"
        class="MagentaA11y__brand--name active"
        href="/home"
      >
        A11y
      </a>
    </div>
    <button
      aria-controls="main"
      aria-expanded="false"
      aria-label="Menu"
      class="MagentaA11y-icon-button"
      data-size="small"
      type="button"
    >
      <svg
        aria-hidden="true"
        focusable="false"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3.50003 17.6344V16.1345H20.5V17.6344H3.50003ZM3.50003 12.7498V11.2499H20.5V12.7498H3.50003ZM3.50003 7.86521V6.36523H20.5V7.86521H3.50003Z"
        />
      </svg>
    </button>
    <nav
      aria-label="main"
      class="MagentaA11y__navbar__nav"
    >
      <ul
        class="MagentaA11y__nav-items"
      >
        <li
          class="MagentaA11y__nav-items--item"
        >
          <a
            class="MagentaA11y__nav-items--link"
            href="/web-criteria/component/overview"
          >
            <span
              class="MagentaA11y__nav-items--label"
            >
              Web Criteria
            </span>
          </a>
        </li>
        <li
          class="MagentaA11y__nav-items--item"
        >
          <a
            class="MagentaA11y__nav-items--link"
            href="/native-criteria/controls/overview"
          >
            <span
              class="MagentaA11y__nav-items--label"
            >
              Native App Criteria
            </span>
          </a>
        </li>
        <li
          class="MagentaA11y__nav-items--item"
        >
          <a
            class="MagentaA11y__nav-items--link"
            href="/about"
          >
            <span
              class="MagentaA11y__nav-items--label"
            >
              About Us
            </span>
          </a>
        </li>
      </ul>
    </nav>
  </div>
</DocumentFragment>
`;
